q2_types/feature_data/_format.py

```diff
@@ -142,142 +143,168 @@ def _validate_(self, level):
     'TSVTaxonomyDirectoryFormat', 'taxonomy.tsv', TSVTaxonomyFormat)
 
 
-class DNAFASTAFormat(model.TextFileFormat):
+class FASTAFormat(model.TextFileFormat):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.aligned = False
+        self.alphabet = None
+
     def _validate_(self, level):
-        FASTADNAValidator = re.compile(r'[ACGTURYKMSWBDHVN]+\r?\n?')
-        ValidationSet = frozenset(('A', 'C', 'G', 'T', 'U', 'R', 'Y', 'K', 'M',
-                                   'S', 'W', 'B', 'D', 'H', 'V', 'N'))
+        FASTAValidator, ValidationSet = _construct_validator_from_alphabet(
+            self.alphabet)
+        self._validate_FASTA(level, FASTAValidator, ValidationSet)
 
-        _validate_DNAFASTAFormats(self, FASTADNAValidator, ValidationSet,
-                                  level)
+    def _validate_FASTA(self, level, FASTAValidator=None, ValidationSet=None):
+        last_line_was_ID = False
+        ids = {}
 
+        seq_len = 0
+        prev_seq_len = 0
+        prev_seq_start_line = 0
 
-DNASequencesDirectoryFormat = model.SingleFileDirectoryFormat(
-    'DNASequencesDirectoryFormat', 'dna-sequences.fasta', DNAFASTAFormat)
+        level_map = {'min': 100, 'max': float('inf')}
+        max_lines = level_map[level]
 
+        with skbio.io.util.open(str(self), encoding='binary') as fh:
+            try:
+                first = fh.read(6)
+                if first[:3] == b'\xEF\xBB\xBF':
+                    first = first[3:]
 
-class PairedDNASequencesDirectoryFormat(model.DirectoryFormat):
-    left_dna_sequences = model.File('left-dna-sequences.fasta',
-                                    format=DNAFASTAFormat)
-    right_dna_sequences = model.File('right-dna-sequences.fasta',
-                                     format=DNAFASTAFormat)
-
+                # Empty files should validate
+                if first.strip() == b'':
+                    return
 
-class AlignedDNAFASTAFormat(model.TextFileFormat):
-    def _validate_(self, level):
-        FASTADNAValidator = re.compile(r'[ACGTURYKMSWBDHVN.-]+\r?\n?')
-        ValidationSet = frozenset(('A', 'C', 'G', 'T', 'U', 'R', 'Y', 'K', 'M',
-                                   'S', 'W', 'B', 'D', 'H', 'V', 'N', '.',
-                                   '-'))
+                if first[0] != ord(b'>'):
+                    raise ValidationError("First line of file is not a valid "
+                                          "description. Descriptions must "
+                                          "start with '>'")
+                fh.seek(0)
+
+                for line_number, line in enumerate(fh, 1):
+                    line = line.strip()
+                    if line_number >= max_lines:
+                        return
+                    line = line.decode('utf-8-sig')
+
+                    if line.startswith('>'):
+                        if FASTAValidator and ValidationSet:
```

Comment:
why invoke FASTAValidator and ValidationSet here? they are not used within this block...

Reply:
We want to skip these checks in the most permissive FASTA case (and the easiest way to check for that is to check whether those validators are not None as would be the case when no alphabet was set).